---
title: "Comparative Genomics Southern Flounder"
output:
  html_notebook:
    code_folding: hide
    df_print: paged
    highlight: kate
    theme: flatly
    toc: yes
  html_document:
    toc: yes
---

```{r load libraries, message=FALSE, warning=FALSE}

knitr::opts_chunk$set(warning = FALSE, message = FALSE)

source("lib/libraries.R")
source("lib/ggplot.R")

```

# Comparative Genomics

Comparative genomics can be used to determine patterns of homology and synteny between the consensus linkage map and fully sequenced fish genomes. Chris Hollenbeck has created an efficient pipeline (`synteny_mapper.pl`) that will blast the mapped loci to fully sequenced fish genomes on a local database hosted on deepthought and determine their relative positions on the corresponding chromosome in order to identify synteny blocks.

# Create Inputfiles

Several input and configuration files are required by the pipeline to specify the loci being blasted, the genomes the linkage map should be compared to and the arguments/settings for BLAST.

## Linkage map (`map_file`)

The linkage map input file (`map_file`) consists of a text file with linkage map information in a tidy format with specifying the locus name, linkage group and position.

For example:

    LOCUS   LG    POS
    locus1  1     0.0
    locus2  1     0.01
    
It is imperative that the locus names in this file correspond to the locus names in the `fasta`-file with the corresponding sequences. For example, onemap adds a `*` in front of each locus name, this needs to be removed before running the synteny pipeline.

```{r format map_file}

# read fullmap data (generated using onemap)
map <- read.table("results/SFL.map",
                header = TRUE, stringsAsFactors = FALSE) %>%
  separate(LOCUS, c("1", "2", "3"), sep = "_") %>%
  mutate(temp = "dDocent") %>%
  select(temp, `2`, `3`, LG, POS) %>%
  unite(LOCUS, 1:3, sep = "_")

# write map file to data folder
write.table(map, "data/SYNTENY/map.tab", 
            quote = FALSE, row.names = FALSE)

```

## Blast configurations (`blast_args`)

The BLAST+ arguments and their associated settings are specified in the `blast_args` configuration file. The pipeline will perform an `blastn`, i.e. a standard nucleotide blast aligning nucleotides to nucleotides. The recommended algorithm is discontiguous megablast, though this can be changed in the configuration file.

```{r format blast_args}

# define blast arguments/settings
Blast_args <- data.frame(Argument = c("num_alignments", "evalue", "num_threads", "lcase_masking", "dust", "task"),
                         Setting = c("10", "1e-10", "10", "yes", "yes", "dc-megablast"))

# write blast_arg file to data folder
write.table(Blast_args, "data/SYNTENY/blast_args.txt", 
            quote = FALSE, row.names = FALSE, col.names = FALSE)

```

The code for the pipeline can be further edited to change `blastn` [parameters](http://www.ncbi.nlm.nih.gov/books/NBK279675/).

## Fasta file with mapped loci (`fasta_file`)

In order to blast the mapped loci their sequences must be specified in `fasta`-format.

The perl script `filter_fasta_by_map.pl` will use the `map.tab` file to filter the `fasta` file for the *de novo* reference used to map reads during the SNP calling process to create a `fasta`-file containing only the mapped loci for the linkage map.

Execute the perl script from within the directory with the *de novo* reference and the perl script (on deepthought).

    cd FLOUNDER/Synteny/
    perl filter_fasta_by_map.pl map.tab reference.fasta
    
This the script will write a `fasta`-file named `filtered.fasta`. Rename file to `mapped.fasta`.

    mv filter.fasta mapped.fasta
    
## Fasta file of additional unmapped loci (`synteny_map_loci`)

An additional `fasta`-file containing known but unmapped loci from that are not part of the linkage map can be specified. For example, a set of genes identified in a gene expression/transcriptomics study on the same species or markers that could not be mapped in a conventional way because they did not segregate in any of the parents.

If provided the pipeline will attempt to map these loci by blasting them to the species genomes, determining if a hit lies within a synteny block and then using the position within that synteny block to identify the (approximate) position within the linkage map.

The additional `fasta`-file is defined using the `synteny_map_loci` setting in the configuration file.

```{r}

# ----- IMPORT INFORMATION ON LOCI TO SYNTENY MAP ------------------------------------

# Williams et al 2003 stress resonse
stress_response <- read.delim("data/SYNTENY/Stress_response_W03.txt",
           header = TRUE, stringsAsFactors = FALSE)

# Williams et al. 2007 gene expression exposed to estradiol
estradiol_response <- read.delim("data/SYNTENY/Estradiol_response_W07.txt",
                              header = TRUE, stringsAsFactors = FALSE)

# Larsen et al. 2007 gene expression salinity differences
salinity_response <- read.delim("data/SYNTENY/Salinity_response_L07.txt",
                                 header = TRUE, stringsAsFactors = FALSE)

# Faciani et al. 2008 gene expression exposed to pollutants (lab and wild populations)
pollutants_response <- read.delim("data/SYNTENY/Pollutants_response_F08.txt",
                                header = TRUE, stringsAsFactors = FALSE)

# microsatellite in linkage map (Japanese flounder)
Japfl_linkmapI <- read.delim("data/SYNTENY/LinkMapI.txt",
                                  header = TRUE, stringsAsFactors = FALSE)

Japfl_linkmapII <- read.delim("data/SYNTENY/LinkMapII.txt",
                             header = TRUE, stringsAsFactors = FALSE)

# ----- DOWNLOAD FASTA FILES -----------------------------------------------------------

# create vector of accession numbers
acc <- unique(c(stress_response$ACCESSION, 
                estradiol_response$ACCESSION, 
                salinity_response$ACCESSION, 
                pollutants_response$ACCESSION,
                Japfl_linkmapI$ACCESSION,
                Japfl_linkmapII$ACCESSION))

acc <- na.omit(acc)

# split into chunks of 100 samples
acc <- split(acc, ceiling(seq_along(acc)/100))

# use alternate function (standard one has issues with scaffolds)
read.GenBank.FASTA <- function(access.nb)
{
  N <- length(access.nb)
  ## if more than 400 sequences, we break down the requests
  nrequest <- N %/% 400 + as.logical(N %% 400)
  X <- character(0)
  for (i in 1:nrequest) {
    a <- (i - 1) * 400 + 1
    b <- 400 * i
    if (i == nrequest) b <- N
    URL <- paste("http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=",
                 paste(access.nb[a:b], collapse = ","),
                 "&rettype=fasta&retmode=text", sep = "")
    X <- c(X, scan(file = URL, what = "", sep = "\n", quiet = TRUE))
  }
  fl <- tempfile()
  cat(X, sep = "\n", file = fl)
  read.FASTA(fl)
}

gen <- read.GenBank.FASTA(acc)

#export as fasta file
write.dna(gen, "data/SYNTENY/flounderloci.fasta", format = "fasta")

# ----- FORMAT IDs -----------------------------------------------------------------

flounderloci_IDs <- read.delim("data/SYNTENY/flounderloci.txt", sep = "|", header = FALSE) %>%
  rename(LOCUS = V2, ACCESSION = V4) %>%
  separate(ACCESSION, c("ACCESSION", "temp"), sep = -3) %>%
  select(LOCUS, ACCESSION)

```

## Create configuration file (`config.txt`)

The specification for the pipeline are listed in a configuration file.

First, the genomes the linkage map should be compared to are specified. Currently there are 9 genomes in the local database on earth:

* Zebrafish, _Danio rerio_ (drer)
* Stickleback, _Gasterosteus aculatus_ (gacu)
* Fugu, _Takifugu rubripes_ (trub)  
* Nile tilapia, _Oreochromis niloticus_ (onil)
* Asian seabass, _Late calcarifer_  (lcal)
* European seabass, _Dicentrarchus labrax_ (dlab)
* Spotten green puffer, _Tetraodon nigroviridis_ (tnig)
* Japanese flounder, _Paralichthys olivaceus_ (poli)
* Tongue sole, _Cynoglossus semilaevis_ (csem)

```{r format config file I}

Species_comp <- data.frame(Species = c("gacu", "tnig", "trub", "onil", "lcal", "dlab", "poli"),
                          Compare = c("TRUE", "TRUE", "TRUE", "TRUE", "TRUE", "TRUE", "TRUE", "TRUE"))

```

Next, the location of the local blast database is defined, and the input files and  minimum number of basepairs needed for a hit to count as a match specified.

```{r format config file II}

# # determine settings
# Synt_settings <- data.frame(Option = c("db_dir", "fasta_file", "map_file", "blast_args", "min_hit_length"),
#                             Setting = c("/home/blast/db", "mapped.fasta", "map.tab", "blast_args.txt", "50")) 

# if additional unmapped files being mapped
Synt_settings <- data.frame(Option = c("db_dir", "fasta_file", "map_file", "synteny_map_loci", "blast_args", "min_hit_length", "error_rate"),
                            Setting = c("/home/blast/db", "mapped.fasta", "map.tab", "flounderloci.fasta", "blast_args.txt", "50", "0.025"))

# write configuration file
con <- file("data/SYNTENY/config.txt", open = "wt")
writeLines("# List of comparison species", con)
write.table(Species_comp, con, 
            append = FALSE, quote = FALSE, sep = " ", col.names = FALSE, row.names = FALSE)
writeLines("# Settings", con)
write.table(Synt_settings, con, 
            append = TRUE, quote = FALSE, sep = " ", col.names = FALSE, row.names = FALSE)
close(con)

```

# Run synteny_mapper

Currently, the local blast database is hosted on deepthought. Create a project folder containing all the necessary input and configuration files, as well as the fasta-filtering script.

After all the necessary files have been assembled `synteny_mapper.pl` can be executed using:

```{bash}

synteny_mapper.pl config.txt

```

The pipeline will blast the sequences of the mapped loci to each species-genome using the discontiguous-megablast algorithm. Matches are considered similar if more than 50bp overlap and have an e-value <10e-10 (or as specified in `config.txt`). Only sequences with a single hit are considered a match, if loci match multiple locations on a given genome they are discarded.

For successfully aligned loci, the pipeline will code the loci as relative positions along the chromosome/linkage group by dividing the starting position by the total chromosome length (in basepairs and the linkage group position by the total linkage group length (in cM).

The pipeline defines synteny blocks as sets of loci on a linkage group that are found in the same order uninterrupted by any other shared locus on the corresponding chromosome. Order mismatches separated by less than 5% of the total length of the LG/chromosome are considered the result of either small-scale local arrangements or ordering errors due to uncertainty in the mapping process and ignored.

# Pipeline output

Synteny mapper will create a set of output files located in a subdirectory (`results_#`); results folders from previous runs will not be overwritten, rather a new results subdirectory will be created. The results subdirectory contains three files and one subfolder with genome-specific output-files for each species genome the linkage group loci were blasted to.

## Main syteny mapping & summaries

`*-report.out.tab` efficiently summarizes the blast results per species genome documented in the alignment report by listing the header lines from the blast report in a tab-deliminated file which identify the locus name, the genbank entry/contig it was aligned to (including accession number), the alignment score (?? what is that score called ??) and the start and end basepairs of the contig it was aligned to.

`*.synteny.stats` summarizes the total number of loci mapped, the total number of blocks and summary statistics for the number of loci per block, and block sizes on the species genome (in bp) and the linkage map (in cM).

`all.blocks.tab` is a tab-delimited file consisting of 12 columns that offer an overview of all the synteny blocks identified for all species genomes the linkage map was compared to. The first column lists the species genome being compared (`COMP_SPECIES`) and the first and last locus from the linkage map forming that block(`START_LOCUS`, `END_LOCUS`), as well as the number of loci contained in that synteny block (`NUM_LOCI`). The final eight columns document the corresponding chromosomes and linkage groups (`COMP_CHR`, `MAP_LG`) and the start and ending coordinates of the corresponding blocks (`COMP_START`, `COMP_END`,	`MAP_LG`, `MAP_START`, `MAP_STOP`), as well as the length of the blocks in basepairs and cM, respectively (`COMP_SIZE`, `MAP_SIZE`).

`*.synteny.tab` follows the same format as `all.blocks.tab`, except that it does not contain a column specifying the species genome that linkage groups are being compared to.

`*-synteny.blocks` consists of one entry per locus giving the LG and Position and the corresponding chromosome and position on the species genome. Loci are grouped by syteny block (separated by a line of dashes).

## other output files

`config.txt` documents the configuration setting used for the `synteny_mapper.pl` run corresponding to the results.

`loci.txt` contains a list of all loci in the linkage map (one locus per line).

`*.report.out` reports the (actual) alignments of the loci from the linkage map that were successfully blasted to the reference genome.

`*.synteny.log` is a log file documenting the process of organizing loci mapped to a species genome into synteny blocks.

`*.synteny.sum` follows a similar format to `all.blocks.tab`, with the difference being that the first column consists of all the loci in a given synteny block (separated by `,`).

# Analyze synteny output

## Compare genome quality

Load genome assembly stats. Three-spined stickleback (gacu) is a chromosome-level high quality genome assembly but the updated stats are not on genbank. Was not able to pull stats from any of the updated papers.

Perform PCA to identify genomes grouping by quality.

```{r}

# load data
assembly_stats <- read.table("data/SYNTENY/genome_assembly_stats.txt",
                             header = TRUE, row.names = 1,
                             stringsAsFactors = FALSE, sep = "\t") %>%
    select(SEQ_LENGTH_TOTAL, PERC_GAP, N_SCAFFOLD, N50_SCAFF, L50_SCAFF)

pca <- dudi.pca(assembly_stats, scale = TRUE, center = TRUE, scannf = FALSE, nf = 10)

read.table("data/SYNTENY/genome_assembly_stats.txt",
                             header = TRUE, row.names = 1,
                             stringsAsFactors = FALSE, sep = "\t")

```

Extract eigenvalues and plot.

```{r, fig.height=3, fig.width=4}

# eigenvalues
eig <- as.data.frame(pca$eig) %>%
    rownames_to_column() %>%
    rename(EIGENVALUE = `pca$eig`,
           PC = rowname) %>%
    mutate(PERCENT = (EIGENVALUE/(sum(EIGENVALUE))*100))

# plot %variance summarize by each PC
ggplot(eig, aes(x = PC, y = PERCENT)) +
    geom_bar(stat = "identity", color = "black", fill = "darkorange") +
    scale_y_continuous(limits = c(0, 100)) +
    labs(x = "PC", y = "% Variance") +
    theme_standard

```

Plot genomes qualities.

```{r, fig.height=4, fig.width=5}

# individuals coordinated normed to squ root of eigenvalue (l1 is normed to 1)
pc <- pca$li

# plot results
ggplot(pc, aes(x = Axis1, y = Axis2, label = rownames(pc))) +
    geom_text(colour = "black", size = 4) +
    geom_hline(yintercept = 0, colour = "darkblue", linetype = "dashed", size = 0.5) +
    geom_vline(xintercept = 0, colour = "darkblue", linetype = "dashed", size = 0.5) +
    labs(x = "PC1", y = "PC2") +
    theme_standard

```

Plot loadings of stats.

```{r, fig.height=4, fig.width=5}

# loadings of variables (normed to 1, i.e. squared values sum to one)
loadings <- pca$c1

# co gives the variablesâ€™ coordinates, normed to the square root of the eigenvalues

ggplot(loadings, aes(x = CS1, y = CS2, label = rownames(loadings))) +
    geom_text(colour = "black", size = 4) +
    geom_segment(aes(x = 0, y = 0, xend = CS1, yend = CS2), 
                 arrow = arrow(length = unit(0.1,"cm")), size = 1, color = "darkorange") +
    geom_hline(yintercept = 0, colour = "darkblue", linetype = "dashed", size = 0.5) +
    geom_vline(xintercept = 0, colour = "darkblue", linetype = "dashed", size = 0.5) +
    scale_x_continuous(limits = c(-0.75, 0.75)) +
    scale_y_continuous(limits = c(-0.75, 0.75)) +
    labs(x = "Loadings 1", y = "Loadings 2") +
    theme_standard

```

```{r, fig.height=4, fig.width=5}

ggplot() +
    geom_text(data = pc, aes(x = Axis1, y = Axis2, label = rownames(pc)),
                             colour = "darkblue", size = 4) +
    geom_text(data = pca$co, aes(x = Comp1, y = Comp2, label = rownames(loadings)),
                             colour = "darkorange", size = 2, vjust = 1) +
    geom_segment(data = pca$co, 
                 aes(x = 0, y = 0, xend = Comp1, yend = Comp2), 
                 arrow = arrow(length = unit(0.1,"cm")), size = 1, color = "darkorange") +
    geom_hline(yintercept = 0, colour = "darkblue", linetype = "dashed", size = 0.5) +
    geom_vline(xintercept = 0, colour = "darkblue", linetype = "dashed", size = 0.5) +
    labs(x = "PC1", y = "PC2") +
    theme_standard

```

## Exploratory analysis

Create function to import `*.synteny.stats` output file.

```{r}

# function to read in *synteny.stats files 
read.synteny.stats <- function(dir, species) {  
  temp <- lapply(species, function(sp){  
    filename <- paste(sp, ".synteny.stats", sep = "")  
    path <- file.path(dir, filename)  
    synteny.stats <- read.table(path, header = FALSE, sep = ":", stringsAsFactors = FALSE,
                                col.names = c("STAT", "species"))  
    synteny.stats <- setNames(data.frame(t(synteny.stats[, -1])), synteny.stats[, 1]) %>%  
      mutate(SPECIES = sp) %>%  
      select(SPECIES, 1:13)  
  }) 
  ldply(temp, rbind) %>% 
    dplyr::rename(N_LOCI = `Total Loci in Blocks`, 
                  N_BLOCKS = `Total Blocks`, 
                  MEAN_LOCI = `Mean loci per block`, 
                  MIN_LOCI = `Min loci per block`, 
                  MAX_LOCI = `Max loci per block`, 
                  TOTAL_BP = `Total Comp Block Size (bp)`, 
                  MEAN_BP = `Mean Comp Block Size (bp)`, 
                  MIN_BP = `Min Comp Block Size (bp)`, 
                  MAX_BP = `Max Comp Block Size (bp)`, 
                  TOTAL_cM = `Total Map Block Size (cM)`, 
                  MEAN_cM = `Mean Map Block Size (cM)`, 
                  MIN_cM = `Min Map Block Size (cM)`, 
                  MAX_cM = `Max Map Block Size (cM)`) 
}   

```

Read in `*.synteny.stats` file and plot overview.

```{r, fig.height=10, fig.width=12}

# Synteny results overview 
species <- c("dlab", "gacu", "lcal", "onil", "poli", "tnig", "trub") 

synteny_stats <- read.synteny.stats("data/SYNTENY", species)

temp <- synteny_stats %>%
    select(-MIN_LOCI, -MAX_LOCI, -MIN_BP, -MAX_BP, -MIN_cM, -MAX_cM) %>%
    mutate(`%MAP` = TOTAL_cM/1605.43)

# # plot synteny mapping results 
# plot.synteny.stats <- function(synteny_stats) { 
#   tidy <- synteny_stats %>% 
#     gather("STAT", "RESULT", 2:14) 
#   ggplot(tidy, aes(x = SPECIES, y = RESULT, fill = SPECIES)) + 
#     geom_bar(color = "black", stat = "identity") + 
#     facet_wrap( ~ STAT, scales = "free") +
#       theme_facet
# } 
# 
# plot.synteny.stats(synteny_stats) 

```

Plot overview total number of loci, number of blocks and total length (cM/bp)

```{r}

# plot overview
plot.synteny.total <- function(synteny_stats) { 
  
  # Format data set
  tidy <- synteny_stats %>% 
    gather("STAT", "RESULT", 2:14) %>%
    filter(STAT %in% c("N_LOCI", "N_BLOCKS", "TOTAL_BP", "TOTAL_cM"))
  
  # plot
  ggplot(tidy, aes(x = SPECIES, y = RESULT, fill = SPECIES)) + 
    geom_bar(stat = "identity",
             color = "black", fill = "darkorange" ) +
    labs(x = "\ncomparison species genome", y = "total number/length\n") +
    facet_wrap(~ STAT, scales = "free") +
    theme_facet
} 

plot.synteny.total(synteny_stats)

```

Compare genome sizes of comparison species.

```{r}

SPECIES <- c("dlab", "gacu", "lcal", "onil", "poli", "tnig", "trub") 
GENOME <- c(578000000, 463000000, 586000000, 657000000, 643000000, 240000000, 281000000)

genome <- data.frame(SPECIES, GENOME) %>%
    arrange(SPECIES)

synteny_stats <- left_join(synteny_stats, genome) %>%
    mutate(GENOME_COV = (TOTAL_BP/GENOME)*100,
           LINKMAP = 1605.43,
           LINKMAP_COV = (TOTAL_cM/LINKMAP)*100,
           LOCI = 2848,
           LOCI_COV = (N_LOCI/LOCI)*100)
# 
# # plot genome size vs coverage
# synteny_stats %>%
#     select(GENOME, GENOME_COV, LINKMAP_COV, LOCI_COV) %>%
#     gather(key = COVERAGE, value = PERCENT, 2:4) %>%
#     ggplot(aes(x = GENOME, y = PERCENT, color = COVERAGE)) +
#     geom_point(shape = 1, size = 2) +
#     geom_smooth(method = lm, se = FALSE, size = 1) +
#     labs(x = "genome size [bp]", y = "% covered") +
#     theme_standard

temp <- synteny_stats %>%
    select(SPECIES, N_BLOCKS, N_LOCI, MEAN_LOCI, TOTAL_BP, MEAN_BP, TOTAL_cM, MEAN_cM, GENOME, GENOME_COV, LINKMAP_COV, LOCI_COV)

write.table(temp, "results/synt.stats", col.names = TRUE, row.names = FALSE, quote = FALSE)

```

Plot proportion of genome and linkage map, respectively covered by sytneny blocks.

```{r}

ggplot(synteny_stats, aes(x = GENOME_COV, y = LINKMAP_COV)) +
    geom_point(shape = 1, size = 3, color = "darkblue") +
    geom_smooth(method = lm, se = TRUE, size = .5, color = "blue") +
    labs(x = "% comparison species genome covered by synt blocks",
         y = "% linkmap covered by synt blocks") +
    theme_standard

```

Plot genome size vs the proportion of loci in linkage blocks

```{r}

ggplot(synteny_stats, aes(x = GENOME, y = LOCI_COV)) +
    geom_point(shape = 1, size = 2, color = "darkblue") +
    geom_smooth(method = lm, se = TRUE, size = .75, color = "blue") +
    labs(x = "genome size",
         y = "% loci in synteny blocks") +
    theme_standard

```

Plot linkage map coverage compared to proportion of loci in synteny blocks

```{r}

ggplot(synteny_stats, aes(x = LINKMAP_COV, y = LOCI_COV)) +
    geom_point(shape = 1, size = 2, color = "darkblue") +
    geom_smooth(method = lm, se = TRUE, size = .75, color = "blue") +
    labs(x = "% linkmap covered by synt blocks",
         y = "% loci in synteny blocks") +
    theme_standard

```

Plot mean number of loci, mean number of blocks, and mean length (cM/bp):

```{r}

# plot overview
plot.synteny.mean <- function(synteny_stats) { 
  
  # Format data set
  tidy <- synteny_stats %>% 
    gather("STAT", "RESULT", 2:18) %>%
    filter(STAT %in% c("N_BLOCKS", "MEAN_LOCI", "MEAN_BP", "MEAN_cM"))
  
  # plot
  ggplot(tidy, aes(x = SPECIES, y = RESULT)) + 
    geom_bar(stat = "identity",
             color = "black", fill = "darkorange" ) +
    labs(x = "\nComparison Species Genome", y = "mean Number/Length\n") +
    facet_wrap(~ STAT, scales = "free") +
    theme_facet
} 

plot.synteny.mean(synteny_stats)

```

Plot overview:

```{r fig.height=3, fig.width=12}

# Format data set
tidy <- synteny_stats %>% 
    gather("STAT", "RESULT", 2:20) %>%
    filter(STAT %in% c("LOCI_COV", "GENOME_COV", "LINKMAP_COV"))

# plot

# library(Cairo)

# Cairo(width = 1000, height = 500, 
#       file="fig/synt_stats.png", type = "png", pointsize  = 12,
#       bg = "transparent", canvas = "white", 
#       units = "px", dpi = "auto")

ggplot(tidy, aes(x = SPECIES, y = RESULT, fill = SPECIES)) + 
    geom_bar(stat = "identity",
             color = "black", size = 0.7, fill = "orange" ) +
    labs(x = "\nComparison Species Genome", y = " ") +
    facet_wrap(~ STAT, scales = "free") +
    theme_facet

# dev.off()

```

Comparison genome size

```{r fig.height=10, fig.width=12}

temp <- synteny_stats %>%
    select(SPECIES, GENOME, N_LOCI, N_BLOCKS, MEAN_LOCI, TOTAL_BP, TOTAL_cM, GENOME_COV, LINKMAP_COV, LOCI_COV) %>%
    gather("STAT", "RESULT", 3:10)

ggplot(temp, aes(x = GENOME, y = RESULT)) +
    geom_point(shape = 1, size = 3, color = "darkblue") +
    geom_smooth(method = lm, se = TRUE, color = "darkblue", size = 0.75) + 
    facet_wrap(~ STAT, scales = "free") +
    labs(x = "genome size [bp]", y = " ") +
    theme_standard

```

Comparison relationships number of blocks:

```{r fig.height=8, fig.width=12}

synteny_stats %>%
    select(SPECIES, N_BLOCKS, N_LOCI, MEAN_LOCI, TOTAL_BP, TOTAL_cM, GENOME_COV, LINKMAP_COV) %>%
    gather("STAT", "RESULT", 3:8) %>%
    ggplot(aes(x = N_BLOCKS, y = RESULT)) +
        geom_point(shape = 1, size = 2) +
        geom_smooth(method = lm, se = TRUE, color = "darkorange", size = 0.5, fill = "grey90") + 
        facet_wrap(~ STAT, scales = "free") +
        labs(x = "number of blocks", y = " ") +
        theme_standard

```

Distribution of synteny block length in genome (bp), linkage map (cM), and number of loci (n).

```{r, fig.height=8, fig.width=6}

# read in synteny block overview
read.all.blocks <- function(dir) { 
  path <- file.path(dir, "all.blocks.tab") 
  synt_blocks <- read.table(path, header = TRUE,
                            colClasses = c("factor", "factor", "factor", "numeric", 
                                           "factor", "integer", "integer", "integer",
                                           "factor", "numeric", "numeric", "numeric"))
}   

synt_blocks <- read.all.blocks("data/SYNTENY")

temp <- synt_blocks %>%
    select(COMP_SPECIES, COMP_SIZE, MAP_SIZE, NUM_LOCI) %>%
    mutate(COMP_SIZE = COMP_SIZE/1000) %>%
    gather(key = COMP, value = LENGTH, 2:4)

ggplot(temp, aes(x = LENGTH)) +
    geom_histogram(color = "black", fill = "darkorange") +
    facet_grid(COMP_SPECIES ~ COMP, scales = "free_x") +
    labs(x = "length (in kb/cM)", y = "number of synteny blocks") +
    theme_facet

View(synt_blocks)

```

## Determine total % of linkage map/genome covered

```{r}

# all syn blocks by linkage groups
blocks <- read.table("data/SYNTENY/all.blocks.tab", 
                         header=TRUE, stringsAsFactors = FALSE) %>%
    select(MAP_LG, MAP_START, MAP_STOP) %>%
    unique() 

# identify blocks where start/end are switched
temp <- filter(blocks, MAP_START > MAP_STOP) %>%
    rename(MAP_STOP = MAP_START,
           MAP_START = MAP_STOP) %>%
    select(MAP_START, MAP_STOP)

# retain only blocks where start < stop
blocks <- filter(blocks, MAP_START <= MAP_STOP)

# add swapped blocks back in
blocks <- bind_rows(blocks, temp) %>%
    arrange(MAP_LG, MAP_START, MAP_STOP)

# create empty list
total_blocks <- list()

# for each linkage group collapse overlapping ranges
for(i in 1:24){
    LG <- filter(blocks, MAP_LG == i) %>%
    arrange(MAP_START, MAP_STOP) %>% 
    group_by(g = cumsum(cummax(lag(MAP_STOP, default = first(MAP_STOP))) < MAP_START)) %>%
    summarise(MAP_START = first(MAP_START), MAP_STOP = max(MAP_STOP)) %>%
    select(-g) %>%
    mutate(LG = i,
           SIZE = MAP_STOP - MAP_START)
    
    total_blocks[[i]] <- LG
}

# merge blocks for all LG
total_blocks <- ldply(total_blocks, data.frame) %>%
    select(LG, MAP_START, MAP_STOP, SIZE)

# total cM of link map covered by synt blocks
total <- sum(total_blocks$SIZE)

# % of link map covered by synt blocks
percent <- (total/1605.43)*100

```

In total (i.e. across all species), `r percent`% of the linkage map is covered by synteny blocks.

## Determine homologous chromosomes

```{r, fig.height=8, fig.width=10}

# read in synteny block overview
read.all.blocks <- function(dir) { 
  path <- file.path(dir, "all.blocks.tab") 
  synt_blocks <- read.table(path, header = TRUE,
                            colClasses = c("factor", "factor", "factor", "numeric", 
                                           "factor", "integer", "integer", "integer",
                                           "factor", "numeric", "numeric", "numeric"))
}   

synt_blocks <- read.all.blocks("data/SYNTENY")

ggplot(synt_blocks, aes(x = COMP_SIZE, y = MAP_SIZE)) +
    geom_point() +
    geom_smooth(method = lm) +
    facet_wrap(~COMP_SPECIES, scales = "free") +
    theme_facet

# write data frame with homologous linkage groups (all species)
LG.Homologs <- function(synteny_blocks){
  synt_blocks %>%
    select(COMP_SPECIES, MAP_LG, COMP_CHR) %>%
    unite(HOMOLOG, 1:3, sep = ">") %>%
    distinct() %>%
    separate(HOMOLOG, into = c("COMP_SPECIES", "MAP_LG", "COMP_CHR"), sep = ">") %>%
    arrange(COMP_SPECIES)
}

Homologs <- LG.Homologs(synt_blocks)

```

## Create oxford bubble plots:

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}

# plot oxford bubble plot
# arguments 
# synt_blocks = data.frame in format all.blocks.tab
# species = vector of species to be plotted

plot.oxford <- function(synt_blocks, species) {

  lapply(species, function(species){
    
    # set factor levels for LG order
    LG <- c("1", "2", "3", "4", "5", "6", "7", "8", 
            "9", "10", "11", "12", "13", "14", "15", "16",
            "17", "18", "19", "20", "21", "22", "23", "24")
    
    # identify order of Chromosomes according to LG order
    CHR <- synt_blocks %>%
      filter(COMP_SPECIES %in% species) %>%
      select(COMP_SPECIES, MAP_LG, COMP_CHR) %>%
      distinct(MAP_LG, .keep_all = TRUE)
    
    CHR <- droplevels(CHR)
    CHR$MAP_LG <- factor(CHR$MAP_LG, levels = LG)
    CHR <- arrange(CHR, MAP_LG)
    
    CHR <- as.character(CHR$COMP_CHR)
    
    # determine number of blocks per chromosome/linkage group combinations
    plot <- synt_blocks %>%
      filter(COMP_SPECIES %in% species) %>%
      select(COMP_SPECIES, MAP_LG, COMP_CHR) %>%
      group_by(MAP_LG, COMP_CHR) %>%
      count() %>%
      filter(!is.na(COMP_CHR))
    
    # order LG and CHR on x and y axis
    plot$COMP_CHR <- factor(plot$COMP_CHR, levels = unique(CHR))
    plot$MAP_LG <- factor(plot$MAP_LG, levels = LG)
    
    # plot LG vs CHR and number of blocks
    ggplot(plot, aes(x = MAP_LG, y = COMP_CHR)) +
      geom_point(aes(size = n), 
                 shape = 21, stroke = 1.1, color = "black", fill = "grey95") +
      scale_size_continuous(range = c(2,15),
                            breaks = c(1, 5, 10)) +
      scale_x_discrete(expand = c(0, 1)) +
      scale_y_discrete(expand = c(0, 1)) +
      scale_fill_discrete("Number of Syntenic Blocks", 
                          breaks=c(1, 5, 10), 
                          labels=c("1", "5", "10")) +
      labs(x = "\nLinkage Group Southern Flounder", y = paste("Chromosome", species, "\n")) +
      theme_classic() +
      theme(
        axis.title = element_text(size = 26),
        axis.title.y = element_text(vjust = 1.5),
        axis.line.x = element_line(color = "black", size = 0.5),
        axis.line.y = element_line(color = "black", size = 0.5),
        axis.ticks.x = element_line(color = "black", size = 1),
        axis.ticks.y = element_line(color = "black", size = 1),
        axis.ticks.length = unit(0.2, "cm"),
        axis.text.x = element_text(size = 18),
        axis.text.y = element_text(size = 18),
        
        legend.position = "bottom",
        legend.text = element_text(size = 18),
        legend.title = element_text(size = 18))
  })
}

plot.oxford(synt_blocks, c("dlab", "lcal", "gacu", "onil", "poli", "tnig", "trub"))

```

# Compare efficiency of synteny mapping loci

```{r, fig.height=3, fig.width=4}

syn_loci <- read.table("data/SYNTENY/all.synteny.mapped.out",
                       header=TRUE, stringsAsFactors = FALSE) %>%
    select(LOCUS, SUPPORT) %>%
    separate(SUPPORT, sep = ",",
             into = c("genome1", "genome2", "genome3", 
                      "genome4", "genome5", "genome6", 
                      "genome7")) %>%
    gather(key = temp, value = SPECIES, 2:8) %>%
    select(-temp) %>%
    filter(!is.na(SPECIES))

# count number of loci supported by each species
per_species <- syn_loci %>%
    group_by(SPECIES) %>%
    summarise(N_LOCI = n())

# count number of loci supported by only one species
per_locus <- syn_loci %>%
    group_by(LOCUS) %>%
    summarise(N_GENOMES = n())

# distribution of number of supporting genomes per locus
ggplot(per_locus, aes(x = N_GENOMES)) +
    geom_histogram(binwidth = 1, color = "black", fill = "orange") +
    theme_standard

unique_loci <- per_locus %>%
    filter(N_GENOMES == 1)

proportion_unique <- (nrow(unique_loci)/3053)*100

# count unique loci per species
unique_loci <- syn_loci %>%
    filter(LOCUS %in% unique_loci$LOCUS)

uniq_per_species <- unique_loci %>%
    group_by(SPECIES) %>%
    summarize(UNIQUE_LOCI = n()) %>%
    mutate(PERCENT_TOTAL = (UNIQUE_LOCI/3053)*100,
           PERCENT_TOTAL_UNIQUE = (UNIQUE_LOCI/1215)*100)

per_species <- left_join(uniq_per_species, per_species) %>%
    mutate(PERCENT_UNIQUE = (UNIQUE_LOCI/N_LOCI)*100,
           PERCENT_LOCI = (N_LOCI/3053)*100) %>%
    select(SPECIES, N_LOCI, PERCENT_LOCI, UNIQUE_LOCI, PERCENT_UNIQUE, PERCENT_TOTAL, PERCENT_TOTAL_UNIQUE)

# synteny blocks w/less than 3 loci

```
