## ------------------------------------------------------------------------- ##
# ---------------------------  import stats.out ----------------------------- #

# import stats.out file generated by haplotyper
# arguments: file (file path as ".../stats.out")

read.hap.stats <- function(file)
{
  read.delim(file, skip = 1, header = TRUE, na.strings = "-",
             colClasses = c("character", "integer", "integer",	"integer",
                            "integer", "numeric", "character",	"integer",
                            "integer",	"integer",	"character"))
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# ---------------------------  import ind_stats.out ------------------------- #

# import stats.out file generated by haplotyper
# arguments: file (file path as ".../ind_stats.out")

read.ind.hap.stats <- function(file)
{
  read.delim(file, header = TRUE, stringsAsFactors = FALSE)
}
## ------------------------------------------------------------------------- ##



## ------------------------------------------------------------------------- ##
# -----------------------  plot filtered vs. passed ------------------------- #

# plot filtered vs passed number of loci
# arguments: hap_stats (data frame with stats.out information)

library(ggplot2)

plot.filter.status <- function(hap_stats)
{
  ggplot(hap_stats, aes(x = Status, stat = "identity")) +
    geom_bar() +
    labs(x = "Filter Status", y = "Number of Loci") +
    theme_standard
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# ------------------------  remove filtered loci ---------------------------- #

# returns data frame containing only loci that passed haplotyping process
# arguments: hap_stats (data frame with stats.out information)
#            x (threshold value)

library(dplyr)

remove.filtered.haps <- function(hap_stats)
{
  hap_stats %>%
    filter(Status == "PASSED") %>%
    select(-Comment, -Status)
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# ---------------------- plot distribution (loci)  -------------------------- #

library(tidyr)
library(ggplot2)
library(ggthemes)

# plot distribution of haplotyper stats
# arguments: hap_stats (data frame with stats.out, including only passed loci)

plot.hap.stats <- function(hap_stats)
{
  
  # format into tidy data set (assumes only passed loci)
  tidy <- hap_stats %>%
    select(-Total_Inds, -Inds_Haplotyped) %>%
    gather("STAT", "LOCI", 2:7)
  
  # plot distributions
  ggplot(tidy, aes(x = LOCI, stat = "bin")) +
    geom_histogram(color = "black", fill = "grey") +
    facet_wrap(~STAT, scales = "free") +
    labs(x = "") +
    theme_classic() +
    theme(
      axis.text = element_text(size = 11),
      axis.title = element_text(size = 16),
      axis.title.y = element_text(vjust = 1.5),
      
      legend.position = "bottom",
      
      panel.background = element_rect(fill = "white", color = NA),
      panel.border = element_rect(fill = NA, color = "black"),
      panel.grid.major = element_line(color = "grey85"),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey95", color = "black"),
      strip.text.x = element_text(size = 16),
      strip.text.y = element_text(size = 16))
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# -------------------- plot distribution (individuals)  --------------------- #

library(tidyr)
library(dplyr)
library(ggplot2)
library(ggthemes)

# plot distribution of haplotyper stats
# arguments: hap_ind_stats (data frame with ind.stats.out)

plot.ind.hap.stats <- function(hap_ind_stats)
{
  
  # format into tidy data set (assumes only passed loci)
  tidy <- hap_ind_stats %>%
    select(-Total_Loci, -Total_Failed) %>%
    gather("STAT", "LOCI", 2:5)
  
  # plot distributions
  ggplot(tidy, aes(x = LOCI, stat = "bin")) +
    geom_histogram(color = "black", fill = "grey") +
    facet_wrap(~STAT, scales = "free") +
    labs(x = "stat") +
    theme_classic() +
    theme(
      axis.text = element_text(size = 11),
      axis.title = element_text(size = 16),
      axis.title.y = element_text(vjust = 1.5),
      
      legend.position = "bottom",
      
      panel.background = element_rect(fill = "white", color = NA),
      panel.border = element_rect(fill = NA, color = "black"),
      panel.grid.major = element_line(color = "grey85"),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey95", color = "black"),
      strip.text.x = element_text(size = 16),
      strip.text.y = element_text(size = 16))
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# ------------------ filter no. potential paralogs  ------------------------- #

# returns vector of loci to remove based on no. of potential paralogs detected
# arguments: hap_stats (data frame with stats.out information)
#            x (threshold value)

filter.loci.paralogs <- function(hap_stats, x)
{
  Poss_Paralog <- hap_stats %>%
    filter(Poss_Paralog > x)
  Poss_Paralog$Locus
}

## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# ------------------ filter no. SNPs per locus  ----------------------------- #

# returns vector of loci to remove based on no. of SNPs per locus
# arguments: hap_stats (data frame with stats.out information)
#            x (threshold value)

filter.loci.noSNPs <- function(hap_stats, x)
{
  NoSNPs <- hap_stats %>%
    filter(Sites > x)
  NoSNPs$Locus
}
## ------------------------------------------------------------------------- ##

## ------------------------------------------------------------------------- ##
# ------------------ filter no. haplotypes per locus  ----------------------- #

# returns vector of loci to remove based on no. of haplotypes per locus
# arguments: hap_stats (data frame with stats.out information)
#            x (threshold value)

filter.loci.NoHaps <- function(hap_stats, x)
{
  NoHaps <- hap_stats %>%
    filter(Haplotypes > x)
  NoHaps$Locus
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# ------------ filter loci low coverage/genotyping error  ------------------- #

# returns vector of loci to remove based on no. ind flagged for low coverage/
# allelic drop out and/or potential genotyping error for given locus

# arguments: hap_stats (data frame with stats.out information)
#            x (threshold value)

filter.loci.geno_err <- function(hap_stats, x)
{
  Geno_Err <- hap_stats %>%
    filter(Low_Cov.Geno_Err > x)
  Geno_Err$Locus
}
## ------------------------------------------------------------------------- ##

## ------------------------------------------------------------------------- ##
# ------------ filter loci missing data   ----------------------------------- #

# returns vector of loci to remove based on genotype call rate

# arguments: hap_stats (data frame with stats.out information)
#            x (threshold value)

filter.loci.prop_haplotyped <- function(hap_stats, x)
{
  Prop_Haplotyped <- hap_stats %>%
    filter(Prop_Haplotyped < x)
  Prop_Haplotyped$Locus
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# -------------------- filter ind possible paralogs  ------------------------ #

# returns vector of individuals to remove based on no. loci flagged as
# possible paralogs for an individual

# arguments: ind_hap_stats (data frame with ind_stats.out information)
#            x (threshold value)

filter.ind.paralogs <- function(ind_hap_stats, x)
{
  Poss_ParalogInd <- ind_hap_stats %>%
    filter(Poss_Paralogs > x)
  Poss_ParalogInd <- Poss_ParalogInd$Ind
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# --------- filter ind proportion successfully haplotyped loci  ------------- #

# returns vector of individuals to remove based on proportion of successfully
# haplotyped loci for a given individual

# arguments: ind_hap_stats (data frame with ind_stats.out information)
#            x (threshold value)

filter.ind.prop_success <- function(ind_hap_stats, x)
{
  SuccessIndv <- ind_hap_stats %>%
    filter(Prop_Success < x)
  SuccessIndv <- SuccessIndv$Ind
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# -------------- filter ind low coverage/genotyping errors  ----------------- #

# returns vector of individuals to remove based on number of loci with low
# coverage/allele drop out and/or potential genotyping errors

# arguments: ind_hap_stats (data frame with ind_stats.out information)
#            x (threshold value)

filter.ind.geno_err <- function(ind_hap_stats, x)
{
  Geno_Err_Indv <- ind_hap_stats %>%
    filter(Low_Coverage.Errors > x)
  Geno_Err_Indv <- Geno_Err_Indv$Ind
}
## ------------------------------------------------------------------------- ##


## ------------------------------------------------------------------------- ##
# ---------------- filter ind high no. missing genotypes  ------------------- #

# returns vector of individuals to remove based on number of missing genotypes

# arguments: ind_hap_stats (data frame with ind_stats.out information)
#            x (threshold value)

filter.ind.miss_geno <- function(ind_hap_stats, x)
{
  MissGenoIndv <- ind_hap_stats %>%
    filter(Miss_Genotype > x)
  MissGenoIndv <- MissGenoIndv$Ind
}
## ------------------------------------------------------------------------- ##